scripts/build/gen_relocate_app.py

```diff
@@ -256,16 +266,20 @@ def print_linker_sections(list_sections):
     return print_string
 
 
+def add_phdr(memory_type, phdrs):
+    return f'{memory_type} {phdrs[memory_type] if memory_type in phdrs else ""}'
```

Comment:
PHDRs are used on the other xtensa boards, not intel_adsp, in fact. I kept support for them because I'm not sure how they are used externally, it might break something on some flasher or so.
That said, I'm not sure how doing this on post processing would be easier: right now, this information is there, in the linker script. Relocation will just place a few more bits in it, allowing the phdrs to be kept (as, inside the linker script, last explicit phdr "wins"). A post processing step would need this map information again, so we'd need to move it. Unless I'm missing something and talking non-sense =D
And yes, the world would be a better place if we could simply get rid of the PHDRS, they seem to be only used in xtensa. I wonder if it's something akin to GNU autotools, and it's just being copied since ever but no-one knows why (and it's not needed anymore).

Reply:
> PHDRs are used on the other xtensa boards, not intel_adsp, in fact.

Just making sure: you mean non-intel_adsp xtensa boards that _do not use rimage_? Examples?

> Frankly this is a bug in rimage (program headers are for runtime linkage!), and we should probably fix it there. But if we have to accommodate buggy signing tools,...

Please don't accommodate a buggy rimage but submit rimage fixes instead if you know what to fix. It's not exactly like rimage has too many picky maintainers, quite the opposite.

I suspect there was some... "vision" of rimage as a stable and loosely coupled tool that you install in your PATH and rarely update but in practice you should alway to use the latest rimage main branch if you don't want to waste a lot of time and that's what most people do thanks to git submodules and now west which maintain a tight coupling.

PS: there is really no reason for `rimage` to be in C and in fact there's an incomplete parser in Python already, it's run every time you build SOF: https://github.com/thesofproject/sof/commit/f8ca8536fd772e55451f6507242ad9465e94c554#diff-f6125b00ced5798fa16565d83deb700969f4f79ea3f6fd32931ede1bdbc0aab1R1492. I digress sorry.

Reply:
> In this PR, [soc/xtensa/sample_controller/include/xtensa-sample-controller.ld](https://github.com/zephyrproject-rtos/zephyr/pull/51466/files#diff-da0e125eec424802ca09dfadccb75d62f2066e36422fb38966c3f99aa427329c). This is the one used for qemu-xtensa, and it has a PHDRS section. This one doesn't seem to use rimage.

I'm not familiar with this but... does a sample matter here? Samples can be changed, no?

> There are also the NXP adsp ones (such as soc/xtensa/nxp_adsp/imx8/linker.ld), but those use rimage.

Very funny coincidence: I just created `NO_RIMAGE_PLATFORMS = ['imx8', 'imx8x', 'imx8m']` in this commit https://github.com/thesofproject/sof/commit/3a9413eebd8260 and another, equivalent list a long time ago in CMake at https://github.com/thesofproject/sof/commit/9818410347b2

@dbaluta is imx8* ever using rimage?

Reply:
@marc-hb IMX uses rimage to create the *ri files. But we don't sign the images.